### PR TITLE
Python 3.7 support

### LINKED ISSUE #1706: Python 3.7 issue because of new re.Pattern
Hi,

after upgrading to Python 3.7, Moto raises the following exception when I run my tests:
`AttributeError: 're.Pattern' object has no attribute 'decode'`

Python Version: 3.7
Moto Version: 1.3.3 

I already looked through the code and I think I found the cause for this (see below).

But first, here is how to reproduce:

```
In [1]: import moto

In [2]: moto.mock_s3_deprecated().start()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-8b514e5bb847> in <module>()
----> 1 moto.mock_s3_deprecated().start()

.../lib/python3.7/site-packages/moto/core/models.py in start(self, reset)
     54                 backend.reset()
     55 
---> 56         self.enable_patching()
     57 
     58     def stop(self):

.../lib/python3.7/site-packages/moto/core/models.py in enable_patching(self)
    113                         method=method,
    114                         uri=re.compile(key),
--> 115                         body=convert_httpretty_response(value),
    116                     )
    117 

.../lib/python3.7/site-packages/moto/packages/httpretty/core.py in register_uri(cls, method, uri, body, adding_headers, forcing_headers, status, responses, match_querystring, **headers)
    962 
    963         matcher = URIMatcher(uri, entries_for_this_uri,
--> 964                              match_querystring)
    965         if matcher in cls._entries:
    966             matcher.entries.extend(cls._entries[matcher])

~/workspace/cnag/cms/cms/__init__.py in patched_init(self, uri, entries, match_querystring)
     13             POTENTIAL_HTTP_PORTS.add(int(result.port or 80))
     14     else:
---> 15         self.info = core.URIInfo.from_uri(uri, entries)
     16 
     17     self.entries = entries

.../lib/python3.7/site-packages/moto/packages/httpretty/core.py in from_uri(cls, uri, entry)
    751     @classmethod
    752     def from_uri(cls, uri, entry):
--> 753         result = urlsplit(uri)
    754         if result.scheme == 'https':
    755             POTENTIAL_HTTPS_PORTS.add(int(result.port or 443))

/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/parse.py in urlsplit(url, scheme, allow_fragments)
    398     Note that we don't break the components up in smaller bits
    399     (e.g. netloc is a single string) and we don't expand % escapes."""
--> 400     url, scheme, _coerce_result = _coerce_args(url, scheme)
    401     allow_fragments = bool(allow_fragments)
    402     key = url, scheme, allow_fragments, type(url), type(scheme)

/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/parse.py in _coerce_args(*args)
    121     if str_input:
    122         return args + (_noop,)
--> 123     return _decode_args(args) + (_encode_result,)
    124 
    125 # Result objects are more helpful than simple tuples

/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/parse.py in _decode_args(args, encoding, errors)
    105 def _decode_args(args, encoding=_implicit_encoding,
    106                        errors=_implicit_errors):
--> 107     return tuple(x.decode(encoding, errors) if x else '' for x in args)
    108 
    109 def _coerce_args(*args):

/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/parse.py in <genexpr>(.0)
    105 def _decode_args(args, encoding=_implicit_encoding,
    106                        errors=_implicit_errors):
--> 107     return tuple(x.decode(encoding, errors) if x else '' for x in args)
    108 
    109 def _coerce_args(*args):

AttributeError: 're.Pattern' object has no attribute 'decode'

```

If I'm not mistaken this happens because the vendored version of `httpretty` (0.8.10) in `moto.packages.httpretty` doesn't work with Python 3.7. It checks for `SRE_Pattern` only but since Python 3.7 introduced `re.Pattern` it doesn't recognize it is a regular expression pattern and treats it as a string. Thus the `decode()` call.

https://github.com/spulec/moto/blob/bb6da93891482ddcb87c23e6abddc0e451f72d30/moto/packages/httpretty/core.py#L775

The latest version of `httpretty` (0.9.5) doesn't have that issue and should work fine.

So I think updating the vendored version of `httpretty` should make the problem go away.

Moto is a great convenience when testing AWS interactions. Thanks for that!

### PR DESCRIPTION
**Pull request by yan12125**
Fixes #1706

`.travis.yml` is so complicated as Python 3.7 is not fully supported on Travis CI. See https://github.com/travis-ci/travis-ci/issues/9815

### CODE REVIEW

**Comment by yan12125** [comment]
All tests finally pass. Networking on Travis CI VMs is not so stable...

File: .travis.yml [hunk 1]
```diff
@@ -8,6 +8,16 @@ python:
 env:
     - TEST_SERVER_MODE=false
     - TEST_SERVER_MODE=true
+matrix:
```
Comment (terrycain) [inline .travis.yml]: Can you not add 3.7 into the `python:` list above?

**Comment by terrycain** [comment]
other than that comment it looks ok

File: .travis.yml (hunk 1 above)
Comment (yan12125) [reply .travis.yml]: It doesn't work - https://travis-ci.org/spulec/moto/builds/407918370.

Maybe I should add a comment to remind future visitors about the relevant Travis CI issue?

Comment (terrycain) [reply .travis.yml]: Wow that's odd. Yeah add a comment, and if you dont mind, poke travis to see if they're aware of that as I'd think that should "just work"

**Comment by terrycain** [comment]
@JackDanger you dont mind testing on 3.7 right?

File: .travis.yml (hunk 1 above)
Comment (yan12125) [reply .travis.yml]: Updated!

Yes they aware - the "sudo + xenial" solution comes from a Travis staff - https://github.com/travis-ci/travis-ci/issues/9815#issuecomment-401756442. Sigh.

**Comment by lalithakolla** [comment]
is there any reason this is not merged yet ?? it would be nice to use python3.7

**Comment by yan12125** [comment]
@lalithakolla: I believe @JackDanger's approval is required.

**Comment by terrycain** [comment]
No so much required more a second opinion would be nice, worse case in a few days I'll merge.

**Comment by yan12125** [comment]
> worse case in a few days I'll merge.

Seems @JackDanger is too busy to check this PR. How about just going ahead?

**Comment by yan12125** [comment]
Any news? I rebased to the latest master and it still looks good.

**Comment by spulec** [comment]
Looks great; thanks!

**Comment by yan12125** [comment]
Thanks! But seems the fix is not included in version 1.3.6?

And here's something strange for version 1.3.6 - the latest commit on https://github.com/spulec/moto/commits/master is 57f9691a52245ffd9d7c0317e90953d054070c65, but the 1.3.6 tag is on https://github.com/spulec/moto/commit/4227cfd72eb245e2a3b7cacbb3334ae24407ac04

**Comment by jonafato** [comment]
Echoing @yan12125's comment here. Any chance of a stable release with 3.7 support soon?

(Thanks for building this!)

**Pull request merged**